File: scala/scala-2.13.8/src/jmh/scala/dev/nulli/jvm_benchmarks/scala_2_13_8/benchmark/ArrayLoopBenchmarks.scala
package dev.nulli.jvm_benchmarks.scala_2_13_8.benchmark

import dev.nulli.jvm_benchmarks.scala_2_13_8.execution.ArrayLoopExecution
import org.openjdk.jmh.annotations.Benchmark
import org.openjdk.jmh.infra.Blackhole

class ArrayLoopBenchmarks {

    @Benchmark
    def forLoop(ex: ArrayLoopExecution, bl: Blackhole): Unit = {
        val modifier = ex.modifier.get
        for (value <- ex.values.get) {
            bl.consume(value + modifier)
        }
    }

    @Benchmark
    def whileLoop(ex: ArrayLoopExecution, bl: Blackhole): Unit = {
        val modifier = ex.modifier.get
        val values = ex.values.get
        var i = 0
        while (i < values.length) {
            bl.consume(values(i) + modifier)
            i += 1
        }
    }

    @Benchmark
    def doWhileLoop(ex: ArrayLoopExecution, bl: Blackhole): Unit = {
        val modifier = ex.modifier.get
        val values = ex.values.get
        var i = 0
        do {
            bl.consume(values(i) + modifier)
            i += 1
        } while (i < values.length)
    }

    @Benchmark
    def forEachLoop(ex: ArrayLoopExecution, bl: Blackhole): Unit = {
        val modifier = ex.modifier.get
        ex.values.get.foreach(value => bl.consume(value + modifier))
    }
}

File: kotlin/kotlin-1.6.10/src/jmh/kotlin/dev/nulli/jvm_benchmarks/kotlin_1_6_10/benchmark/ArrayLoopBenchmarks.kt
package dev.nulli.jvm_benchmarks.kotlin_1_6_10.benchmark

import dev.nulli.jvm_benchmarks.kotlin_1_6_10.execution.ArrayLoopExecution
import org.openjdk.jmh.annotations.Benchmark
import org.openjdk.jmh.infra.Blackhole

open class ArrayLoopBenchmarks {

    @Benchmark
    fun forLoop(ex: ArrayLoopExecution, bl: Blackhole) {
        val modifier = ex.modifier!!
        for (value in ex.values!!) {
            bl.consume(value + modifier)
        }
    }

    @Benchmark
    fun forIndexLoop(ex: ArrayLoopExecution, bl: Blackhole) {
        val values = ex.values!!
        val modifier = ex.modifier!!
        for (i in values.indices) {
            bl.consume(values[i] + modifier)
        }
    }

    @Benchmark
    fun whileLoop(ex: ArrayLoopExecution, bl: Blackhole) {
        val values = ex.values!!
        val modifier = ex.modifier!!
        var i = 0
        while (i < values.size) {
            bl.consume(values[i++] + modifier)
        }
    }

    @Benchmark
    fun doWhileLoop(ex: ArrayLoopExecution, bl: Blackhole) {
        val values = ex.values!!
        val modifier = ex.modifier!!
        var i = 0
        do {
            bl.consume(values[i++] + modifier)
        } while (i < values.size)
    }

    @Benchmark
    fun forEach(ex: ArrayLoopExecution, bl: Blackhole) {
        val modifier = ex.modifier!!
        ex.values!!.forEach { value -> bl.consume(value + modifier) }
    }
}
